zebra/zebra_mlag.c

```diff
@@ -23,16 +23,557 @@
+ */
+void zebra_mlag_send_deregister(void)
+{
+	struct stream *s = NULL;
+
+	s = stream_new(sizeof(struct mlag_msg));
+
+	stream_putl(s, MLAG_DEREGISTER);
+	stream_putw(s, MLAG_MSG_NULL_PAYLOAD);
+	stream_putw(s, MLAG_MSG_NO_BATCH);
+	stream_fifo_push_safe(zrouter.mlag_info.mlag_fifo, s);
+	zebra_mlag_signal_write_thread();
+
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug("%s: Enqueued MLAG De-Register to MLAG Thread ",
+			   __func__);
+}
+
+/*
+ * API To handle MLAG Received data
+ * Decodes the data using protobuf and enqueue to main thread
+ * main thread publish this to clients based on client subscription
+ */
+void zebra_mlag_process_mlag_data(uint8_t *data, uint32_t len)
+{
+	struct stream *s = NULL;
+	struct stream *s1 = NULL;
+	int msg_type = 0;
+
+	s = stream_new(ZEBRA_MAX_PACKET_SIZ);
+	msg_type = zebra_mlag_protobuf_decode_message(s, data, len);
+
+	if (msg_type <= 0) {
+		/* Something went wrong in decoding */
+		stream_free(s);
+		zlog_err("%s: failed to process mlag data-%d, %u", __func__,
+			 msg_type, len);
+		return;
+	}
+
+	/*
+	 * additional four bytes are for message type
+	 */
+	s1 = stream_new(stream_get_endp(s) + ZEBRA_MLAG_METADATA_LEN);
+	stream_putl(s1, msg_type);
+	stream_put(s1, s->data, stream_get_endp(s));
+	thread_add_event(zrouter.master, zebra_mlag_post_data_from_main_thread,
+			 s1, 0, NULL);
+	stream_free(s);
+}
+
+/**********************End of MLAG Interaction********************************/
+
+/************************MLAG Thread Processing*******************************/
+
+/*
+ * after posting every 'ZEBRA_MLAG_POST_LIMIT' packets, MLAG Thread will be
+ * yielded to give CPU for other threads
+ */
+#define ZEBRA_MLAG_POST_LIMIT 100
+
+/*
+ * This thread reads the clients data from the Global queue and encodes with
+ * protobuf and pass on to the MLAG socket.
+ */
+static int zebra_mlag_client_msg_handler(struct thread *event)
+{
+	struct stream *s;
+	uint32_t wr_count = 0;
+	uint32_t msg_type = 0;
+	uint32_t max_count = 0;
+	int len = 0;
+
+	wr_count = stream_fifo_count_safe(zrouter.mlag_info.mlag_fifo);
+	if (IS_ZEBRA_DEBUG_MLAG)
+		zlog_debug(":%s: Processing MLAG write, %u messages in queue",
+			   __func__, wr_count);
+
+	max_count = MIN(wr_count, ZEBRA_MLAG_POST_LIMIT);
+
+	for (wr_count = 0; wr_count < max_count; wr_count++) {
+		s = stream_fifo_pop_safe(zrouter.mlag_info.mlag_fifo);
+		if (!s) {
+			zlog_debug(":%s: Got a NULL Messages, some thing wrong",
+				   __func__);
+			break;
+		}
+
+		/*
+		 * Encode the data now
+		 */
+		len = zebra_mlag_protobuf_encode_client_data(s, &msg_type);
+
+		/*
+		 * write to MCLAGD
+		 */
+		if (len > 0) {
+			zebra_mlag_private_write_data(mlag_wr_buffer, len);
+
+			/*
+			 * If message type is De-register, send a signal to main
+			 * thread, so that necessary cleanup will be done by
+			 * main thread.
+			 */
+			if (msg_type == MLAG_DEREGISTER) {
+				thread_add_event(zrouter.master,
+						 zebra_mlag_terminate_pthread,
+						 NULL, 0, NULL);
```

Comment:
at this point, a terminate event is queued for zebra main pthread:
1) this call doesn't use a back-reference to ensure the event is unique, so ... is there any way it could be invoked more than once?
2) are there any paths where subsequent messages or event have been enqueued to the mlag pthread that can cause errors or crashes as they're now racing with the main pthread?

Reply:
In case by this time, Registration is enqueued again, main thread will not Terminate the thread. below check will make sure this.


static int zebra_mlag_terminate_pthread(struct thread *event)
{
        if (IS_ZEBRA_DEBUG_MLAG)
                zlog_debug("Zebra MLAG write thread terminate called");

        if (zrouter.mlag_info.clients_interested_cnt) {
                if (IS_ZEBRA_DEBUG_MLAG)
                        zlog_debug(
                                "Zebra MLAG: still some clients are interested");
                return 0;
        }